lib/teleterm/services/clientcache/clientcache.go

```diff
@@ -0,0 +1,182 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package clientcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// Cache stores clients keyed by cluster URI.
+// Safe for concurrent access.
+// Closes all clients and wipes the cache on Clear.
+type Cache struct {
+	clusters.Resolver
+	mu  sync.Mutex
+	log *logrus.Entry
+	// clients keep mapping between cluster URI
+	// (both root and leaf) and proxy clients
+	clients map[uri.ResourceURI]*client.ProxyClient
+	// group prevents duplicate requests to create clients
+	// for a given cluster URI
+	group singleflight.Group
+}
+
+// Config describes the client cache configuration.
+type Config struct {
+	clusters.Resolver
+	Log *logrus.Entry
+}
+
+// New creates an instance of Cache.
+func New(c Config) *Cache {
+	return &Cache{
+		log:      c.Log,
+		clients:  make(map[uri.ResourceURI]*client.ProxyClient),
+		Resolver: c.Resolver,
+	}
+}
+
+// Get returns a client from the cache if there is one,
+// otherwise it dials the remote server.
+// The caller should not close the returned client.
+func (c *Cache) Get(ctx context.Context, clusterURI uri.ResourceURI) (*client.ProxyClient, error) {
+	groupClt, err, _ := c.group.Do(clusterURI.String(), func() (any, error) {
+		if fromCache := c.getFromCache(clusterURI); fromCache != nil {
+			return fromCache, nil
+		}
+
+		_, clusterClient, err := c.ResolveCluster(clusterURI)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		//nolint:staticcheck // SA1019. TODO(gzdunek): Update to use client.ClusterClient.
+		newProxyClient, err := clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// We'll save the client in the cache, so we don't have to
+		// build a new connection next time.
+		// All cached clients will be closed when the daemon exits.
+		if err := c.addToCache(clusterURI, newProxyClient); err != nil {
+			return nil, trace.NewAggregate(err, newProxyClient.Close())
+		}
+
+		c.log.WithField("cluster", clusterURI).Info("Added client to cache.")
+
+		return newProxyClient, nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clt, ok := groupClt.(*client.ProxyClient)
+	if !ok {
+		return nil, trace.BadParameter("unexpected type %T received for proxy client", groupClt)
+	}
+
+	return clt, nil
+}
+
+// InvalidateForRootCluster closes and removes clients from the cache
+// for the root cluster and its leaf clusters.
+func (c *Cache) InvalidateForRootCluster(rootClusterURI uri.ResourceURI) error {
```

Comment:
```suggestion
func (c *Cache) ClearForRootCluster(rootClusterURI uri.ResourceURI) error {
```

idk what the conventions are in Go, but this makes it play better with the existing `Clear` function I suppose.